### PR TITLE
frontend/feature: #227 Driver-Alternate-Number

### PR DESCRIPTION
**Pull request by mohdshoaib18169**
mber

## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
 Added Alternate Number Flow in drivers code, Added Alternate Number Button on homescreen of drivers app , resolve a little bug in OTP modal of homecreen

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Frontend/ui-driver/src/Components/InAppKeyboardModal/View.purs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                    , fontStyle state.headingConfig.fontStyle
+                    , margin state.headingConfig.margin
+                    , visibility state.headingConfig.visibility
+                    , cornerRadius state.headingConfig.cornerRadius
+                    , padding state.headingConfig.padding
+                    , weight state.headingConfig.weight
+                    ]
+                    
+                ]
+              , linearLayout
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Components/InAppKeyboardModal/View.purs:94]: create a separate view function for this layout. it seems very long

File: Frontend/ui-driver/src/Flow.purs [hunk 2]
```diff
@@ -452,6 +461,106 @@ driverProfileFlow = do
...
+                else do
+                  modifyScreenState $ DriverDetailsScreenStateType (\driverDetailsScreen ->updatedState)
+                  driverDetailsFlow
+            Left errorPayload -> do
+               let alternateNumber = (if updatedState.props.isEditAlternateMobile then updatedState.data.driverAlternateMobile else Nothing)
+               if (errorPayload.code == 400 && (decodeErrorMessage errorPayload.response.errorMessage) == "Person with this mobile number already exists") then do
+                modifyScreenState $ DriverDetailsScreenStateType $ \driverDetailsScreen -> updatedState { props {numberExistError = true,keyboardModalType = ALTERNATE_NUMBER }}
+                driverDetailsFlow
+               else do
+                 -- pure $ toast $ (decodeErrorCode errorPayload.response.errorCode)
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Flow.purs:483]: remove commented lines

File: Frontend/ui-driver/src/Screens/Types.purs [hunk 3]
```diff
@@ -446,17 +447,33 @@ type DriverDetailsScreenState = {
   props :: DriverDetailsScreenStateProps
 }
 
+data KeyboardModalType = ALTERNATE_NUMBER | OTP | NONE
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Screens/Types.purs:450]: instead oh `ALTERNATE_NUMBER` make it `MOBILE_NUMBER` which makes it more generic.

File: Frontend/ui-driver/src/Services/Backend.purs [hunk 4]
```diff
@@ -658,3 +658,62 @@ getPerformanceBT payload = do
...
+
+makeVerifyAlternateNumberOtpRequest :: String -> DriverAlternateNumberOtpReq
+makeVerifyAlternateNumberOtpRequest otp = DriverAlternateNumberOtpReq {
+
+      "otp" : otp
+ }
+
+-----------------------------------removeAlternateNumber-----------------------------------
+
+--removeAlternateNumberBT :: RemoveAlternateNumberRequest -> FlowBT String RemoveAlternateNumberResp
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Services/Backend.purs:710]: remove commented lines from all place.

**Review by Vignesh-772 — CHANGES_REQUESTED** [review]

**Comment by Rohit4997** [comment]
@mohdshoaib18169  Can you rebase this PR

**Comment by Rohit4997** [comment]
> @mohdshoaib18169 Can you rebase this PR

@mohdshoaib18169  can you also add video attachment here ?

File: Frontend/ui-driver/src/Screens/HomeScreen/Handler.purs [hunk 5]
```diff
@@ -78,6 +80,13 @@ homeScreen = do
       modifyScreenState $ HomeScreenStateType (\homeScreen → state)
       App.BackT $ App.BackPoint <$> (pure $ UPDATE_STAGE stage)
     GoToNotifications -> App.BackT $ App.BackPoint <$> pure GO_TO_NOTIFICATIONS
+    AddAlternateNumber state -> do 
+      modifyScreenState $ DriverDetailsScreenStateType (\driverDetailsScreen -> driverDetailsScreen { data 
+      { driverAlternateMobile = Nothing
+        }, 
+        props  { isEditAlternateMobile = false, keyboardModalType = MOBILE__NUMBER}})
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Screens/HomeScreen/Handler.purs:87]: change the state in Controller itself

File: Frontend/ui-driver/src/Screens/HomeScreen/View.purs [hunk 6]
```diff
@@ -302,28 +303,38 @@ driverStatus push state =
...
-    pattern : "[0-9]*,4",
-    fontSize : 24,
-    focusIndex : state.props.enterOtpFocusIndex,
-    otpIncorrect : if (state.props.otpAttemptsExceeded) then false else (state.props.otpIncorrect),
-    otpAttemptsExceeded : (state.props.otpAttemptsExceeded)
-  })
+  InAppKeyboardModal.view (push <<< InAppKeyboardModalAction) (enterOtpState state)
+
+enterOtpState :: HomeScreenState -> InAppKeyboardModal.InAppKeyboardModalState
+enterOtpState state = let
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Screens/HomeScreen/View.purs:309]: move config to componentConfig

**Review by Vignesh-772 — CHANGES_REQUESTED** [review]

File: Frontend/ui-driver/src/Components/InAppKeyboardModal/View.purs [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+      , weight state.inputTextConfig.weight
+      , gravity state.inputTextConfig.gravity
+      , visibility state.inputTextConfig.visibility
+      , cornerRadius state.inputTextConfig.cornerRadius
+      , padding state.inputTextConfig.padding
+      , margin state.inputTextConfig.margin
+      , onClick push (const (OnclickTextBox 0))
+      ]]
+
+otp :: forall w . (Action -> Effect Unit) -> InAppKeyboardModalState -> PrestoDOM (Effect Unit) w
```
[the remark this answers is not in the document]
Comment (mohdshoaib18169) [reply Frontend/ui-driver/src/Components/InAppKeyboardModal/View.purs]: changed bro

File: Frontend/ui-driver/src/Flow.purs [hunk 8]
```diff
@@ -462,6 +471,101 @@ driverProfileFlow = do
...
+    RESEND_ALTERNATE_OTP updatedState -> do
+      let number =  (if (updatedState.props.isEditAlternateMobile) then updatedState.data.driverEditAlternateMobile else updatedState.data.driverAlternateMobile)
+      getAlternateMobileResendOtpResp <- lift $ lift $ Remote.resendAlternateNumberOTP (makeResendAlternateNumberOtpRequest (fromMaybe "" (number)))
+      case getAlternateMobileResendOtpResp of
+            Right (AlternateNumberResendOTPResp resp) -> do
+                pure $ toast (getString OTP_RESENT)
+                modifyScreenState $ DriverDetailsScreenStateType (\driverDetailsScreen -> updatedState)
+                driverDetailsFlow
+            Left errorPayload -> do
+              modifyScreenState $ DriverDetailsScreenStateType (\driverDetailsScreen -> updatedState)
```
[the remark this answers is not in the document]
Comment (mohdshoaib18169) [reply Frontend/ui-driver/src/Flow.purs]: done

File: Frontend/ui-driver/src/Flow.purs [hunk 9]
```diff
@@ -462,6 +471,101 @@ driverProfileFlow = do
...
+                  pure $ toast (getString OTP_RESEND_LIMIT_EXCEEDED)
+                  driverDetailsFlow
+              else do
+                  pure $ toast (decodeErrorMessage errorPayload.response.errorMessage)
+                  modifyScreenState $ DriverDetailsScreenStateType (\driverDetailsScreen -> updatedState {data{ driverAlternateMobile = (if(updatedState.props.isEditAlternateMobile) then updatedState.data.driverAlternateMobile else Nothing), driverEditAlternateMobile = Nothing} , props{  otpIncorrect = false ,otpAttemptsExceeded = false ,keyboardModalType = NONE , alternateMobileOtp = "",checkAlternateNumber =(updatedState.props.isEditAlternateMobile == false) }})
+                  driverDetailsFlow
+
+
+    VERIFY_OTP state -> do
+       let toast_value = (if (state.props.isEditAlternateMobile == false) then (getString NUMBER_ADDED_SUCCESSFULLY) else (getString NUMBER_EDITED_SUCCESSFULLY))
```
[the remark this answers is not in the document]
Comment (mohdshoaib18169) [reply Frontend/ui-driver/src/Flow.purs]: done

File: Frontend/ui-driver/src/Resource/Localizable/Language/KN.purs [hunk 10]
```diff
@@ -353,3 +353,28 @@ getKN script = case script of
   FOR_UPDATES_SEE_ALERTS -> "ನವೀಕರಣಗಳಿಗಾಗಿ ಘೋಷಣೆಗಳನ್ನು ನೋಡಿ"
   SHARE_OPTIONS -> "ಹಂಚಿಕೆ ಆಯ್ಕೆಗಳು"
   ENTER_PASSWORD -> "ಪಾಸ್ವರ್ಡ್ ನಮೂದಿಸಿ"
+  --APP_NOT_WORKING -> "ಅಪ್ಲಿಕೇಶನ್ ಕಾರ್ಯನಿರ್ವಹಿಸುತ್ತಿಲ್ಲ"
```
[the remark this answers is not in the document]
Comment (mohdshoaib18169) [reply Frontend/ui-driver/src/Resource/Localizable/Language/KN.purs]: done

File: Frontend/ui-driver/src/Screens/DriverDetailsScreen/Controller.purs [hunk 11]
```diff
@@ -41,18 +53,52 @@ instance loggableAction :: Loggable Action where
...
+    PopUpModalAction (PopUpModal.OnButton1Click) -> trackAppActionClick appId (getScreen DRIVER_DETAILS_SCREEN) "pop_up_modal" "On_click_button1"
+    PopUpModalAction (PopUpModal.OnButton2Click) -> trackAppActionClick appId (getScreen DRIVER_DETAILS_SCREEN) "pop_up_modal" "On_click_button2"
+    PopUpModalActions (PopUpModal.OnButton2Click) -> trackAppActionClick appId (getScreen DRIVER_DETAILS_SCREEN) "pop_up_modal2" "On_click_button2"
+    InAppKeyboardModalOtp (InAppKeyboardModal.OnClickResendOtp) -> trackAppActionClick appId (getScreen DRIVER_DETAILS_SCREEN) "in_app_otp_modal" "on_click_done"
+    _ -> trackAppActionClick appId (getScreen DRIVER_DETAILS_SCREEN) "in_app_otp_modal" "on_click_done"
+
 
       
-data ScreenOutput = GoBack DriverDetailsScreenState
+data ScreenOutput = GoBack DriverDetailsScreenState | ValidateAlternateNumber DriverDetailsScreenState | ResendAlternateNumberOTP DriverDetailsScreenState | VerifyAlternateNumberOTP DriverDetailsScreenState | RemoveAlternateNumber DriverDetailsScreenState | GoToHomeScreen DriverDetailsScreenState
```
[the remark this answers is not in the document]
Comment (mohdshoaib18169) [reply Frontend/ui-driver/src/Screens/DriverDetailsScreen/Controller.purs]: done

File: Frontend/ui-driver/src/Screens/DriverDetailsScreen/View.purs [hunk 12]
```diff
@@ -190,23 +216,238 @@ driverDetailsView state =
...
+        , color Color.blue900
+        , margin (Margin 0 8 0 20)
+        ]
+        ]
+
+removeAlternateNumber :: forall w . (Action -> Effect Unit) -> ST.DriverDetailsScreenState -> PrestoDOM (Effect Unit) w
+removeAlternateNumber push state =
+  PopUpModal.view (push <<< PopUpModalAction) (removeAlternateNumberConfig state)
 
+removeAlternateNumberConfig :: ST.DriverDetailsScreenState -> PopUpModalConfig.Config
```
[the remark this answers is not in the document]
Comment (mohdshoaib18169) [reply Frontend/ui-driver/src/Screens/DriverDetailsScreen/View.purs]: done

File: Frontend/ui-driver/src/Screens/HomeScreen/Handler.purs (hunk 5 above)
Comment (mohdshoaib18169) [reply Frontend/ui-driver/src/Screens/HomeScreen/Handler.purs]: did in flow.purs

File: Frontend/ui-driver/src/Screens/HomeScreen/View.purs (hunk 6 above)
Comment (mohdshoaib18169) [reply Frontend/ui-driver/src/Screens/HomeScreen/View.purs]: done

**Comment by witcher-shailesh** [comment]
Duplicate PR, changes already merged.

**Pull request closed without merging**